### ISSUE TITLE
RFC - GitLab 15.0 Deprecations for the Terraform Provider

### ISSUE THREAD

**Issue by PatrickRice-KSC** [gitlabhq/terraform-provider-gitlab#1014]
### Feature Description

Investigate which deprecations we need to account for within the terraform provider that will be removed within GitLab 15.0. The list of deprecations for 15.0 are available here: https://docs.gitlab.com/ee/update/deprecations.html

Initial List (Please feel free to leave comments and I'll maintain the list of ones here, along with whether we've fixed them with a PR)

- [x] Remove test coverage regex from the `gitlab_project` resource [Reference](https://docs.gitlab.com/ee/user/admin_area/settings/continuous_integration.html#required-pipeline-configuration)
- [x] GitLab runner will return `stale` instead of `offline` or `not_connected` [Reference](https://docs.gitlab.com/ee/update/deprecations.html#api-stale-status-returned-instead-of-offline-or-not_connected). [Similar reference for the `never_connected` value](https://docs.gitlab.com/ee/update/deprecations.html#runner-status-not_connected-api-value)
- [x] Legal Approval Status values will no longer be returned by the License Compliance API [Reference](https://docs.gitlab.com/ee/update/deprecations.html#legacy-approval-status-names-from-license-compliance-api)

Deprecations Listed on the page that do *not* impact 15.0 (as they are deferred to 16.0)

- Manual milestone management is no longer supported. There is no resource for this, but I seem to recall seeing a request for one, we should not implement this. [Reference](https://docs.gitlab.com/ee/update/deprecations.html#manual-iteration-management)
- Runner API is removing `active` and `paused` as valid values. [Reference](https://docs.gitlab.com/ee/update/deprecations.html#rest-api-runner-will-not-accept-status-filter-values-of-active-or-paused). [GraphQL has the Same Change](https://docs.gitlab.com/ee/update/deprecations.html#rest-api-runner-will-not-accept-status-filter-values-of-active-or-paused).
- Rest API for Group Runners will no longer accept `project_type` as a value for the `type` argument. [Reference](https://docs.gitlab.com/ee/update/deprecations.html#rest-api-runner-will-not-accept-status-filter-values-of-active-or-paused)
- Runner statuses will no longer return `active` or `paused` for a status [Reference](https://docs.gitlab.com/ee/update/deprecations.html#rest-and-graphql-api-runner-status-will-not-return-paused)

**Comment by PatrickRice-KSC** [comment]
I'm pleasantly surprised by how few deprecation policies really affect us. 

I would propose we make the following updates (and I can create the PRs to do so if we're in alignment):
- For removing the test_coverage_regex, we add a deprecation notice now, and add a check that ignores the value if the remote GitLab version is 15.0 or greater, then plan a breaking change in the 14.0 version that removes the attribute.
- For the GitLab runner change, we don't have a current data source or resource for this API (I have an open PR actually working on this), and as long as I account for the correct status, I think this will update properly for folks when they run the tf provider after we migrate.

Thoughts from the community?

**Comment by Shocktrooper** [comment]
@PatrickRice-KSC Sounds good to me.  Also I think we might need to align the API for Project Approval Rules too. https://docs.gitlab.com/ee/api/merge_request_approvals.html#create-project-level-rule

**Comment by PatrickRice-KSC** [comment]
> @PatrickRice-KSC Sounds good to me. Also I think we might need to align the API for Project Approval Rules too. https://docs.gitlab.com/ee/api/merge_request_approvals.html#create-project-level-rule

Was there a change there that I missed other than deprecating the vulnerability check? If so, I can add that to the list above! I thought the only thing changing there was the vulnerability check, and we don't have a resource for that right now (I want to add one, because that's something that beekeep and I want to use too 🙂 ). That'll be contingent on the GraphQL RFC which I still need to create so we have a consolidated strategy for dealing with GraphQL endpoints.

**Comment by Shocktrooper** [comment]
The deprecation was related to a ticket I submitted awhile ago actually and I closed it out because of the deprecations for 15.0.  Looking at it now the only thing I think that needs to be updated is `report_type` which I don't think is even currently supported for the MR approvals resource we have right now.

**Comment by PatrickRice-KSC** [comment]
Closing this issue, I think we've got all the deprecations handled. The different return values should port automatically when someone runs `apply` for the first time, as the read API will return the new values.
